### PR TITLE
Make retry parameters configurable

### PR DESCRIPTION
**Pull request by valiparsa**
"PIPELINE_CLONE_INTERVAL" and "PIPELINE_CLONE_TIMEOUT" parameters
need to be configurable and those parameters should be possible to
be passed from stewardci-core and also from backend components.
These parameters should have default acceptable values if they are
not set as environment variables.

### CODE REVIEW

File: jenkinsfile-runner-steward-image/scripts/run.sh [hunk 1]
```diff
@@ -23,6 +23,8 @@ declare -r PARAM_VARS_OPTIONAL=(
   'PIPELINE_LOG_FLUENTD_HOST'
   'PIPELINE_LOG_FLUENTD_PORT'
   'PIPELINE_LOG_FLUENTD_TAG'
+  'PIPELINE_CLONE_INTERVAL'
+  'PIPELINE_CLONE_TIMEOUT'
```
Comment (anbrsap) [inline jenkinsfile-runner-steward-image/scripts/run.sh:27]:
```suggestion
  'PIPELINE_CLONE_RETRY_INTERVAL_SEC'
  'PIPELINE_CLONE_RETRY_TIMEOUT_SEC'
```

File: jenkinsfile-runner-steward-image/scripts/run.sh [hunk 2]
```diff
@@ -192,6 +194,24 @@ function with_retries() {
   return "$rc"
 }
 
+function check_retry_params() {
```
Comment (anbrsap) [inline jenkinsfile-runner-steward-image/scripts/run.sh:197]: I don't see the need to implement this as command execution wrapper. The parameter validation should be included in function `with_retries` generically (without referring to the `PIPELINE_CLONE_*` env var names in error messages).

File: jenkinsfile-runner-steward-image/scripts/run.sh [hunk 3]
```diff
@@ -192,6 +194,24 @@ function with_retries() {
...
+  local rc \
+    cmd=("$@") \
+    retry_interval=$2 \
+    timeout_seconds=$3 \
+
+  if ! [[ "$retry_interval" =~ ^[0-9]+$ ]] || ! [[ "$timeout_seconds" =~ ^[0-9]+$ ]]; then
+    printf """\nOnly integer values are allowed for PIPELINE_CLONE_INTERVAL \
+and PIPELINE_CLONE_TIMEOUT parameters but '%s' and '%s' are passed respectively ...\n""" "$retry_interval" "$timeout_seconds" >&2
```
Comment (anbrsap) [inline jenkinsfile-runner-steward-image/scripts/run.sh:205]: Bash doesn't support triple quotes like Python, but a quoted string can span multiple lines:
```suggestion
    printf "\nOnly integer values are allowed for PIPELINE_CLONE_INTERVAL \
and PIPELINE_CLONE_TIMEOUT parameters but '%s' and '%s' are passed respectively ...\n" "$retry_interval" "$timeout_seconds" >&2
```

File: jenkinsfile-runner-steward-image/scripts/run.sh [hunk 4]
```diff
@@ -192,6 +194,24 @@ function with_retries() {
   return "$rc"
 }
 
+function check_retry_params() {
+  local rc \
+    cmd=("$@") \
+    retry_interval=$2 \
+    timeout_seconds=$3 \
+
+  if ! [[ "$retry_interval" =~ ^[0-9]+$ ]] || ! [[ "$timeout_seconds" =~ ^[0-9]+$ ]]; then
```
Comment (anbrsap) [inline jenkinsfile-runner-steward-image/scripts/run.sh:203]: Check each parameter individually and report only those which are invalid, or even the first invalid only.

File: jenkinsfile-runner-steward-image/scripts/run.sh [hunk 5]
```diff
@@ -192,6 +194,24 @@ function with_retries() {
...
+    printf """\nOnly integer values are allowed for PIPELINE_CLONE_INTERVAL \
+and PIPELINE_CLONE_TIMEOUT parameters but '%s' and '%s' are passed respectively ...\n""" "$retry_interval" "$timeout_seconds" >&2
+    return 1
+  elif (( retry_interval > timeout_seconds )); then
+    echo "Parameter PIPELINE_CLONE_INTERVAL cannot be smaller than PIPELINE_CLONE_TIMEOUT in value." >&2
+    return 1
```
Comment (anbrsap) [inline jenkinsfile-runner-steward-image/scripts/run.sh:209]: I would leave this out. According to the implementation of `with_retries`, the timeout is checked _before_ applying the retry delay. So the delay _can_ be longer than the timeout and one single retry could be made if the command fails within the timeout period.

Anyway, the message is wrong.

**Review by anbrsap — CHANGES_REQUESTED** [review]

File: jenkinsfile-runner-steward-image/scripts/run.sh (hunk 2 above)
Comment (valiparsa) [reply jenkinsfile-runner-steward-image/scripts/run.sh:197]: okay i'll do as suggested

File: jenkinsfile-runner-steward-image/scripts/run.sh (hunk 4 above)
Comment (valiparsa) [reply jenkinsfile-runner-steward-image/scripts/run.sh:203]: I'll do as suggested

**Comment by valiparsa** [comment]
Result of test execution:
1- Start a pipelineRun resource while cloudci-git-mirror pod is down:
```
Cloning pipeline repository <git-mirror pipeline repo URL>
Cloning into '.'...
fatal: unable to access '<git-mirror pipeline repo URL>': Failed to connect to <git-mirror pipeline repo URL> port 80: Operation timed out

Retrying with a delay of 15 seconds (130/180 seconds elapsed)...
Cloning into '.'...
fatal: unable to access '<git-mirror pipeline repo URL>': Failed to connect to <git-mirror pipeline repo URL> port 80: Operation timed out

Not retrying anymore as timeout of 180 seconds is reached.

```

2- Start pipelineRun resource when cloudci-git-mirror pod is down and start the pod after few seconds:
```
Cloning pipeline repository <git-mirror pipeline repo URL>
Cloning into '.'...
fatal: unable to access '<git-mirror pipeline repo URL>': Failed to connect to <git-mirror pipeline repo URL> port 80: Operation timed out

Retrying with a delay of 15 seconds (130/180 seconds elapsed)...
Cloning into '.'...
Checking out pipeline from revision 3.0
Already on '3.0'
Your branch is up to date with 'origin/3.0'.
.
.
.
```

In the first test cloudci-git-mirror pod was completely down and the pipeline took 2x130 seconds to fail.

In the second test cloudci-git-mirror pod was initially down but got recovered and the pipeline could be cloned in the second try (after 130+15 seconds).

File: jenkinsfile-runner-steward-image/scripts/run.sh [hunk 6]
```diff
@@ -23,12 +23,18 @@ declare -r PARAM_VARS_OPTIONAL=(
...
+  'PIPELINE_CLONE_RETRY_INTERVAL_SEC'
+  'PIPELINE_CLONE_RETRY_TIMEOUT_SEC'
   'JOB_NAME'
   'RUN_NUMBER'
   'RUN_CAUSE'
   'TERMINATION_LOG_PATH'
 )
 
+declare -r \
+  DEFAULT_PIPELINE_CLONE_RETRY_INTERVAL_SEC=15 \
```
Comment (syurevich) [inline jenkinsfile-runner-steward-image/scripts/run.sh:35]: Is there any reason behind taking exactly 15 sec?

**Review by rinckm — APPROVED** [review]

File: jenkinsfile-runner-steward-image/scripts/run.sh (hunk 6 above)
Comment (valiparsa) [reply jenkinsfile-runner-steward-image/scripts/run.sh:35]: No actually I increased it as we decided to increase from 30 seconds timeout into 180 seconds. If you think we should have 5 seconds interval with 180 seconds timeout I'm fully fine with it

Comment (syurevich) [reply jenkinsfile-runner-steward-image/scripts/run.sh:35]: I think it is ok. It has pros and cons. Let's keep 15 sec.

Comment (valiparsa) [reply jenkinsfile-runner-steward-image/scripts/run.sh:35]: Anyhow we made it configurable in a way these values can be changed in stewardci-core or in backend systems later on

Comment (valiparsa) [reply jenkinsfile-runner-steward-image/scripts/run.sh:35]: https://github.com/SAP/stewardci-core/pull/209

**Pull request merged**